setup.sh

```diff
@@ -9,38 +9,124 @@ function debug() {
     fi
 }
 
+# will attempt to use a package manager to install the missing program, will still stop the script
+# but it should be installed
+function installProg() {
+    progName=$1
+
+    echo "You are missing '$progName', attempting to install...."
+
+
+    # try to install using apt-get
+    if command -v apt-get &>/dev/null; then
+        echo "Using 'apt-get' to install '$progName'"
+
+        sudo apt-get update
+        sudo apt-get install -y "$progName"
+        exit 1
+    fi
+
+    # try to install using apt-get
+    if command -v dnf &>/dev/null; then
+        echo "Using 'dnf' to install '$progName'"
+
+        sudo dnf install -y "$progName"
+        exit 1
+    fi
+
+    # try to install using homebrew
+    if  command -v brew &>/dev/null; then
+        echo "Using 'homebrew' to install '$progName'"
+
+        brew install "$progName"
+        exit 1
+    fi
+
+    # try to install using macports
+    if  command -v port &>/dev/null; then
+        echo "Using 'port' to install '$progName'"
+
+        sudo port install "$progName"
+        exit 1
+    fi
+
+    printf "\e[31mCould not find a package manager. $progName is not installed. Please install it manually!\e[0m\n"
+    exit 1
+}
+
 if ! command -v python3 &>/dev/null; then
-    echo 'Python 3 is not installed. Please install it!'
+    installProg 'python3'
+
+    printf "\e[31mRe-run this script to try again\e[0m\n"
```

Comment:
This doesn't explain why a user might have to try again. I can foresee people complaining about this and making our lives very sad.

Reply:
Actually, I think this line isn't reachable? `installProg` always exits?